More rust training content

brson opened the pull request:
It's been a long time since I've submitted anything, so here's my current branch.

The main things to note here is that project 1 has more text now, and that I've removed gRPC and tokio  from the plan (from discussion with Liu we agreed they were too ambitious for these 5 projects and could come in a later project).

[Rendered](https://brson.github.io/talent-training/rust/index.html)

Tomorrow I will create an issue outlining the next steps so that others might help.

PTAL @overvenus @nolouch

brson commented:
cc @nrc happy to hear any suggestions.

brson commented:
I've pushed another commit that completes the text for project 1.

sticnarf commented on rust/CONTRIBUTING.md:
```diff
@@ -0,0 +1,89 @@
+solutions - students should learn where to get the answers from the
+pre-requisites and lessons.
+
+Proect text may include inline links to pages that offer explanatios of terms
```
Typo.

```suggestion
Project text may include inline links to pages that offer explanations of terms
```

sticnarf commented on rust/CONTRIBUTING.md:
```diff
@@ -0,0 +1,89 @@
+TODO organize this
+
+Every md file has a html relative symlink to index.html:
```
Better to explain the reason? I first felt confused until I find out we are loading the md dynamically according to the url.

sticnarf commented on rust/CONTRIBUTING.md:
```diff
@@ -0,0 +1,89 @@
+
+## Contributor notes
+
+- each md file needs a relative symlink to index.html
```
To be more clarified, let's use `an html relative symlink` same as the beginning.

sticnarf commented on rust/CONTRIBUTING.md:
```diff
@@ -0,0 +1,89 @@
+
+> _Use [crates.io](https://crates.io) to find the documentation
+for the `clap` crate, and implement the command line interface
+such that the `cli_*` test cases pass._
```
On the rendered page, the `_` in the inline code interrupts the `em` block. It's actually a [bug](https://github.com/markedjs/marked/issues/1410) of marked. Should we use html tag as a workaround or wait for them fixing it?

sticnarf commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +234,199 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+
+_Use [crates.io](https://crates.io) to find the documentation
+for the `clap` crate, and implement the command line interface
+such that the `cli_*` test cases pass._
```
The same.

sticnarf commented on rust/plan.md:
```diff
@@ -24,209 +24,243 @@ A suggested workflow:
+**Goals**:
 
-- Lesson: Build-time Rust
+- Use hyper for synchronous networking
```
To my knowledge, as HTTP servers, the latest hyper and iron are all asynchronous and run over the tokio runtime. Maybe we can consider use `http` crate to build a synchronous HTTP server ourselves?

sticnarf replied:
My fault. Only at the master branch iron uses hyper 0.12 but the released 0.6.0 version uses the synchronous hyper 0.10.

sticnarf replied:
Then, how will we make it single threaded first and change it to multi-threaded in the next project? By using hyper 0.10 or iron, that's simply to set the thread num of the pool to 1 first and then make it larger. This does not make much sense.

brson replied:
I was under the impression hyper still offered a synchronous API, but it does not seem to.

Building a synchronous HTTP server could be interesting, requiring direct use of the std networking APIs. It's also demystifying, shows that HTTP is approachable. It's not all that practical, since most people should not implement HTTP themselves.

Another solution would be to use gRPC instead of HTTP.

I'm going to leave this an open question.

sticnarf commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +234,199 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
-that you are not currently working on fixing.
+that you have not yet fixed.
 
 **Question D**: even if a given test suite doesn't contain any tests, why
```
```suggestion
**Question D**: Even if a given test suite doesn't contain any tests, why
```

sticnarf commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +234,199 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+
+Again, the interface for the CLI is:
+
+- `kvs set [KEY] [VALUE]`
```
I think KEY and VALUE are all required arguments. Is it better to change the square brackets to angle brackets `<>`?

brson replied:
Yes, good catch.

sticnarf commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +234,199 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+
+  Set the value of a string key to a string
+
+- `kvs get [KEY]`
```
same

06kellyjac commented on rust/.gitignore:
```diff
@@ -1,2 +1,2 @@
 *~
-yarn-error.log
+target/
```
No newline at eof
```suggestion
target/

```

sticnarf commented on rust/projects/tools/project.md:
```diff
@@ -1,25 +1,120 @@
+
+  Print the version
+
+The `kvs` library contains a type, `KvStore`, that supports the following
```
The behavior is not clearly specified, for example, the behavior of getting the value of a non-existing key.

brson replied:
I'll change the return value of `get` to be `Option<String>`.

sticnarf commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +234,199 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+that you have not yet fixed.
 
 **Question D**: even if a given test suite doesn't contain any tests, why
 might we not want them to run? Besides issuing the above command, how could
```
Do you mean "why might we want to to run?" here

brson replied:
No, I mean "not", but I can see how the sentence suggests the opposite. I'll change it to just "Why might we not want to run empty test suites". The only answer I'm thinking of has to do with compile times. This question may be too obscure to keep.

sticnarf replied:
I still feel the question a bit confusing. Can we add something to lead readers to thinking about the compile time? For example, add "It takes no time to run an empty test.` to the beginning?

sticnarf commented on rust/projects/tools/tests/tests.rs:
```diff
@@ -0,0 +1,8 @@
+#[test]
+fn cli_no_args() {}
```
I will fill the tests.

brson commented on rust/README.md:
```diff
@@ -3,16 +3,14 @@
-- Asyncronous programming with [futures] and [tokio]
-- High-performance networking with [GRPC] and [prost]
+- Asyncronous programming with [futures]
+- Networking with [hyper]
```
I wasn't planning to do anything HTTP2 specific, though I assume hyper will negotiate a HTTP2 upgrade if it can. Do you have any suggestions?

I'm torn about using HTTP for this, since gRPC is what PingCAP uses. But @siddontang and I discussed that it may be too much to require gRPC for this project. If we don't do gRPC in the initial set of projects, I'm thinking it could be in a subsequent project.

brson replied:
@nrc Somewhere @sticnarf suggested having the student just implement basic HTTP for that project, which does force them to write "real" networking code (vs just dropping in hyper).

brson commented on rust/notes.md:
```diff
@@ -37,6 +37,9 @@ in the README.
 - DSTs
+- configuring clippy / rustfmt
+- scripting clippy / rustfmt for CI
+- CI setup
```
Offhand I'd prefer to be opinionated and pick either travis or circle, whichever has the most open source mindshare (or maybe whichever pingcap uses). That way we can be specific about requirements, and have a better chance at automatically verifying they set up CI correctly.

It's not a strong opinion though.

It's also a chance to talk about the README and other boilerplate files, the various badges and other services Rust projects typically use.

brson commented on rust/projects/tools/src/lib.rs:
```diff
@@ -0,0 +1,3 @@
+pub fn main() {
```
Good feedback. I'll change it to `run`, but let me know if you've got a better name. Or better ideas about how to test the cli (I'd rather stay within `cargo test` than explain how some custom scripts work).

brson commented:
Thanks so much for the reviews @sticnarf, @06kellyjac, @overvenus ! I've pushed a commit that addresses all of the feedback except for the problem that there's no synchronous hyper.

brson commented on rust/plan.md:
```diff
@@ -24,209 +24,240 @@ A suggested workflow:
-  - Topics: formatting tips, derive Debug in depth, how does format! work?, log
-    demo, slog and structured logging, env_logger, APIs that take format
-    strings,
+### [Lesson: Formatting, println et. al, log, and slog][t-fmt] ([slides][s-fmt])
```
The motivation is to teach interesting intermediate Rust material that is related to the project at hand. This project has evolved such that there's not a lot of formatting going on though.

I've heard quite a bit of feedback about the scope of the project and the lessons not focusing on only the information needed for the projects.

I'm just going to delete all the lessons and focus on the projects for now.

brson commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +235,198 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+things to do before it is a polished piece of Rust software, ready for
+contributions or publication.
+
+First, public items should have doc comments.
```
I'll add a qualifier.

brson commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +235,198 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+the `target/doc` folder. Note though that `target/doc` folder does not contain
+an `index.html`. Your crate's documentation will be located at
+`target/doc/kvs/index.html`. You can launch a web browser at that location with
+`cargo doc --open`.
```
I don't know. It doesn't help complete the project. The only way I can think to fit it in is as an aside, like "Relatedly, when you need to browse Rust's own documentation, you can run `rustup doc`."

brson commented:
> I think the questions would make good exam questions, but are perhaps not so good as exercises in lessons since there is usually just one right answer, and finding that answer probably won't help the student learn too much.

When I was writing the project, any time I hit a task that could be done multiple ways, with different pros and cons; or when I had some thoughts about the subject beyond the task at hand, I wrote it down as a question, with the intent that the student thinks deeper about the subject and also gains that knowledge.

Some of @sticnarf's comments on his own followup PR probably make two of the questions moot, so I'll probably just drop the notion of questions altogether.

brson commented on rust/projects/tools/project.md:
```diff
@@ -114,19 +235,198 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+information gathered from the type signature. They explain why and how one would
+use a function, what the return value is on both success and failure, error and
+panic conditions. The library you have written is very simple so the
+documentation can be simple as well.
```
Will do.

brson commented on rust/projects/tools/project.md:
````diff
@@ -114,19 +235,198 @@ this case the tests in `tests.rs` (`--test tests`). We can even narrow our testi
+
+```
+rustup component add clippy
+rustup component add rustfix
````
Good catch.

brson commented:
> My other big picture thing is that I worry about the overall goals of the course and how focussed it should be: is this an intermediate course which happens to use networking as a teaching tool? Or is it a networking course? Or a Rust networking course? etc. I think it would be good to think specifically about what audience we expect, and what 'problem' we are solving for them.

Hi @nrc.

This question of the goals of the course, and the intended audience are distressing. It's been asked a number of times, and I've tried to convey it clearly in the readme.

Right now the readme contains this statement:

> Goals of this course are to convey knowledge relevant to real-world Rust programming, including inherited wisdom and best practices; to look "under the hood" and understand how and why Rust works like it does; to tour the world of Rust documentation and code such that the student can learn how to find answers to their questions on their own.

And this

> This book is not for novice programmers, but it is for novice Rust programmers.

> This course is for programmers with some knowledge of data structures, algorithms and I/O.

> This course is for programmers who have begun learning Rust, but still need guidance in production Rust programming and best practices.

> This is a course for active learners — it will lead you in the right direction, but the journey is yours.

> Those who complete the course will have the knowledge and experience to begin to write high performance, reliable, networked applications in Rust.

> This course, combined with Deep Dive TiKV is intended to be enough to enable programmers to meaningfully contribute to TiKV or any other Rust project. It is most specifically designed to teach those in the Chinese Rust community enough Rust to work on TiKV. The language used is intended to be simple so that those who read only a little English can follow.

The goal statement expresses my teaching interests - I want to convey knowledge that can't be found in the books, the types of things that I find myself considering when writing Rust software. The projects - to me - are simply a backdrop for stirring up subjects to dig deep into.

This doesn't seem to resonate with others though, since I hear lots of confusion about the lesson subjects not being relevant to the projects.

So it looks like I need to adjust my idea for this course, focus on projects that teach the technologies that TiKV uses, providing the background material needed for that, and cut everything else out.

I'll think of new statements about the goals and audience, but I'm very glad to take suggestions. Maybe once we have a revised goal statement I or someone else will annotate every topic to explain the motivation, since that is a question that comes up on every pull request.

cc @sticnarf this thread has discussion about the goals of the project.

I've pushed a commit to address @nrc's suggestions, except for mentioning `rustup doc`, which I'm not convinced about. The larger questions of scope can be addressed later.

For the sake of progress I'm going to merge this, but please file bugs if you disagree with any of it.

nrc commented:
I'll keep discussion of general goals on https://github.com/pingcap/talent-plan/issues/24

> This question of the goals of the course, and the intended audience are distressing.

Sorry, I don't mean to cause distress. A contributing factor may be that I have heard several different motivations for the course from different people at different times and I am taking that into account as well as the actual text.

> For the sake of progress I'm going to merge this, but please file bugs if you disagree with any of it.

:+1:
